trulens/nn/models/keras_utils.py

```diff
@@ -52,56 +55,71 @@ def recurse_outputs(obj, prefix=None):
             for i, elem in enumerate(obj):
                 ret.update(recurse_outputs(elem, prefix + [i]))
         else:
-            print(type(obj))
+            tru_logger.warning(type(obj))
         return ret
 
-    def get_arg_path(layer, arg):
+    def get_arg_path(layers, obj):
         """Given the tensor argument path, retrieve the actual tensor(s) in the matching data structure"""
-        if B.is_tensor(arg):
-            return layer_outputs[layer.name][hash_tensor(arg)]
-        elif isinstance(arg, collections.abc.Mapping):
-            return {
-                key: layer_outputs[layer.name][hash_tensor(a)]
-                for key, a in arg.items()
-            }
-        elif isinstance(arg, collections.abc.Iterable):
-            return [layer_outputs[layer.name][hash_tensor(a)] for a in arg]
+
+        def path_from_tensor(a):
+            for layer in layers:
+                try:
+                    return layer_outputs[layer.name][hash_tensor(a)]
+                except KeyError:
+                    continue
+            raise ValueError(
+                f"Unable to find matching layer in {layers}for arg {a}"
+            )
+
+        if B.is_tensor(obj):
+            return path_from_tensor(obj)
+        elif isinstance(obj, collections.abc.Mapping):
+            return {key: path_from_tensor(a) for key, a in obj.items()}
+        elif isinstance(obj, collections.abc.Iterable):
+            return [path_from_tensor(a) for a in obj]
         else:
             raise ValueError("Invalid argument found")
 
     layer_outputs = {}
     layer_input_paths = {}
 
     # Outputs of each layer
-    # TODO: this may be cleaner with node traversal instead of layer traversal
-    for layer in model.layers:
-        layer_outputs[layer.name] = recurse_outputs(layer.output, [layer.name])
-
-    # Path to inputs of each layer
-    for layer in model.layers:
-        out_nodes = set(layer._outbound_nodes)
-        for out_node in out_nodes:
-            next_layer = out_node.outbound_layer
-            args = out_node.call_args
-            kwargs = out_node.call_kwargs
-
-            arg_paths = [get_arg_path(layer, arg) for arg in args]
-            kwarg_paths = {
-                key: get_arg_path(layer, arg) for key, arg in kwargs.items()
-            }
-
-            if next_layer.name not in layer_input_paths:
-                layer_input_paths[next_layer.name] = {
-                    "args": arg_paths,
-                    "kwargs": kwarg_paths
+    node_depths = sorted(list(model._nodes_by_depth.keys()), reverse=True)
```

Comment:
Using model nodes here since each layer can have multiple output nodes if used in different models.